noria/src/data.rs

```diff
@@ -430,16 +430,25 @@ impl<'a> From<&'a DataType> for Cow<'a, str> {
     }
 }
 
-impl From<&'_ DataType> for String {
-    fn from(data: &'_ DataType) -> Self {
-        let cow: Cow<'_, str> = data.into();
-        cow.to_string()
-    }
-}
-
 impl From<DataType> for String {
     fn from(data: DataType) -> Self {
-        (&data).into()
+        match data {
+            DataType::Text(ref s) => s.to_str().unwrap().to_string(),
```

Comment:
I think we should be able to avoid an allocation + copy here?

Reply:
How about `DataType::Text(ref s) => String::from_utf8(s.to_bytes().to_vec()).unwrap()`?

`.to_bytes()` returns a `&[u8]`. In the end, a `String` needs a `Vec`, so there must be one allocation at some point. I don't think there's any other way to build a `String` from an `ArcCStr`.

Reply:
But why are you taking `ref s` when you own the `DataType`? Just take ownership of the `s`.

Reply:
Ohh, I see why you're having problems. I forgot that `String` is an `ArcCStr`, which isn't owned in the first place. Blah. This makes me think that the only way to really provide this is to have `impl From<&'a DataType> for &'a str`. And then leave it to the caller to take ownership if they truly wish.

Reply:
I added the said `&str` implementation, which requires no allocation + copy. The caveat is that where the `Cow<'a, str>` conversion would use `lossy` methods, ours does not (and unwraps, expecting no UTF-8 errors). I'm not sure how we want to handle errors in that case: should we change `From` to a `TryForm`, for instance? Can we rather assume that a string stored in a `DataType` is safe to use?

Reply:
Yeah, I think we only ever build one from a valid UTF-8 string, though you'd have to check the constructors.

Reply:
I have verified all the ways in `data.rs` that a `Text` or `TinyText` variants can be generated :
* `From<Literal>` or `From<&'a Literal>`: `nom_sql::Literal::String` holds a `String`, which must be UTF-8 per Rust docs.
* `TryFrom<&'a [u8]>`: will return an error if the byte-slice is invalid.
* `From<String>` or `From<&'a str>`: both use the previous `TryFrom` — they will panic if the given `String` or `&str` is invalid (which is impossible, because `String` and `&str` are UTF-8 per Rust docs).
* `TryFrom<mysql_common::value::Value>`: will return an error if the byte-slice is invalid.
* `deep_clone(&self)`, or `From<&'a DataType>`: [insert proof by induction 😛]

Thus, we can assume that `Text` and `TinyText` variants are UTF-8, thus no need for a `TryFrom`. Are there other things I should consider, or is it good to merge into master?

Reply:
Awesome! Yes, should be good then. I think personally I would like to see a comment added to the `impl From for &str` that outlines why it's okay to unwrap in the code below (basically a condensed version of the above comment), but that's not _necessary_. Thanks for sticking with it!